hibernate-core/src/main/java/org/hibernate/type/format/StringJsonDocumentReader.java

```diff
@@ -0,0 +1,505 @@
+/*
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.type.format;
+
+import org.hibernate.type.descriptor.WrapperOptions;
+import org.hibernate.type.descriptor.java.JavaType;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.nio.CharBuffer;
+import java.util.NoSuchElementException;
+
+/**
+ * Implementation of <code>JsonDocumentReader</code> for String representation of JSON objects.
+ */
+public class StringJsonDocumentReader extends StringJsonDocument implements  JsonDocumentReader {
+
+	private final CharBuffer json;
```

Comment:
I would very much appreciate if you could use a `String` and int variables for position/limit instead of the `CharBuffer`, as that enables the use of the `JavaType#fromEncodedString` and other special parsing methods that avoid allocating intermediate strings. See what `org.hibernate.dialect.JsonHelper#fromString(org.hibernate.type.descriptor.java.JavaType<?>, org.hibernate.type.descriptor.java.JavaType<?>, org.hibernate.type.descriptor.jdbc.JdbcType, java.lang.String, int, int, boolean, org.hibernate.type.descriptor.WrapperOptions)` did before.